### ISSUE TITLE
Adding new strings through upload doesn't work (JSON)

### ISSUE THREAD

**Issue by jgoclawski** [WeblateOrg/weblate#5942]
I am unable to add new keys to JSON-format component by uploading the file. Our project uses the internal VCS and uploads/downloads the files through API.

I initially stumbled upon this error in a production component with "JSON nested structure" format, but was able to reproduce it on a fresh install with "JSON" format as well.
I have the same error when uploading through API or through GUI. The logs are clear, the file in VCS is not updated on disk. No changes are pending to be commited.

Right now, the only workaround I found is to use the "replace" strategy, which is not optimal for our workflow.

**I already tried**

- Initially I thought it's similar to #5160. But the issue was closed and confirmed as fixed.
- I was able to reproduce the issue on a fresh install, with a minimal example.
- [ x] I've read and searched [the docs](https://docs.weblate.org/) and did not find the answer there.
      If you didn’t try already, try to search there what you wrote above.

**To Reproduce the issue**

1. Have a fresh install (happens on version 4.6 and 4.5.3), no addons, no customizations.
2. Create a new component through file upload - [here's an example zip file](https://drive.google.com/file/d/1tcFnYmFWTo--03ZsOeb0a2mvVpiRI9So/view?usp=sharing). Use the default, autodiscovered settings.
3. Add a new key "newOne" - [here's an example en.json file](https://drive.google.com/file/d/1JAC6JPipSym5jfAnTe0TWo_YRgVuARu3/view?usp=sharing). 
4. Go to your new component, pick English language, go to "Files" -> "Upload translation"
5. Use the example file, select upload mode as "Add new strings"
4. Hit "Upload".
5. Notice how only 9 strings are still present, the new one is not added.

The GUI says "Processed 9 strings from the uploaded files (skipped: 9, not found: 0, updated: 0)." at the top.
The logs say nothing, even with DEBUG mode and log level enabled:
```
uwsgi stderr | [2021-04-29 18:18:27,657: INFO/385] socialwifi/reproduce_test: running source checks for 0 strings
uwsgi stderr | [pid: 385|app: 0|req: 12/36] 172.17.42.1 () {52 vars in 1129 bytes} [Thu Apr 29 18:18:27 2021] POST /upload/socialwifi/reproduce_test/en/ => generated 0 bytes in 41 msecs (HTTP/1.1 302) 8 headers in 391 bytes (1 switches on core 0)
nginx stdout | 172.17.42.1 - - [29/Apr/2021:18:18:27 +0200] "POST /upload/socialwifi/reproduce_test/en/ HTTP/1.1" 302 5 "http://0.0.0.0:8888/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.182 Safari/537.36"
uwsgi stderr | [pid: 385|app: 0|req: 13/37] 172.17.42.1 () {46 vars in 910 bytes} [Thu Apr 29 18:18:27 2021] GET /projects/socialwifi/reproduce_test/en/ => generated 97062 bytes in 542 msecs (HTTP/1.1 200) 9 headers in 449 bytes (1 switches on core 0)
nginx stdout | 172.17.42.1 - - [29/Apr/2021:18:18:28 +0200] "GET /projects/socialwifi/reproduce_test/en/ HTTP/1.1" 200 18054 "http://0.0.0.0:8888/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.182 Safari/537.36"
```

**Expected behavior**

The source language should have the new key added. It doesn't.

**Screenshots**

**Exception traceback**

**Server configuration and status**

Weblate installation:Docker

**Weblate deploy checks**

**Additional context**

**Comment by nijel** [comment]
Skipped should indicate that the string is already present. Are you sure it's not there? It doesn't have to be committed to the file yet.

**Comment by jgoclawski** [comment]
@nijel Hey, thanks for super fast reply, I really appreciate it :)

The initial file in the initial ZIP upload when creating the component has 9 keys and the one being uploaded after - 10 keys.

Initial en.json:
```
{
  "copy": "Copy",
  "back": "Back",
  "cancel": "Cancel",
  "delete": "Delete",
  "edit": "Edit",
  "reset": "Reset",
  "submit": "Submit",
  "activate": "Activate",
  "deactivate": "Deactivate"
}
```

The one uploaded after (notice 4th line - "newOne" key):
```
{
  "copy": "Copy",
  "back": "Back",
  "newOne": "testing",
  "cancel": "Cancel",
  "delete": "Delete",
  "edit": "Edit",
  "reset": "Reset",
  "submit": "Submit",
  "activate": "Activate",
  "deactivate": "Deactivate"
}
```

So I assumed that it should work. It even works when I chose the "replace" file strategy, only the "add new strings" strategy doesn't work.

**Comment by jgoclawski** [comment]
@nijel Thanks! I was just hacking around to try and find the root cause when you commited the fix.
I've applied the patch locally and it works for adding new keys!

Bad news is that it breaks when a source phrase changes. I'm not sure if it's realted to this change or not, so i can create a separate issue if needed.
In the [docs](https://docs.weblate.org/en/weblate-4.6/user/files.html#import-methods) it says that the "add" method "skips the one which already exist" but I think that it failed on that.

Existing file:
```
{
  "one": "One",
  "two": "Two",
}
```

Uploaded file with "add" method:

```
{
  "one": "One",
  "two": "Something else",
}
```
This causes the following error in GUI:
```
File upload has failed: duplicate key value violates unique constraint "trans_unit_translation_id_id_hash_ab945589_uniq" DETAIL: Key (translation_id, id_hash)=(33, 2023870350614736476) already exists.
```
and in logs:
```
uwsgi stderr | [2021-04-30 14:17:09,521: WARNING/2363] Upload error: IntegrityError: duplicate key value violates unique constraint "trans_unit_translation_id_id_hash_ab945589_uniq"
uwsgi stderr | DETAIL:  Key (translation_id, id_hash)=(41, -2330725867778400320) already exists.
uwsgi stderr | 
uwsgi stderr | [pid: 2363|app: 0|req: 65/201] 172.17.42.1 () {64 vars in 1419 bytes} [Fri Apr 30 14:17:09 2021] POST /upload/socialwifi/reproduce_test_2/en/ => generated 0 bytes in 45 msecs (HTTP/1.1 302) 8 headers in 393 bytes (1 switches on core 0)
```

**Comment by nijel** [comment]
That should be fixed by 054bda78f9127a014682fde37d97e3e2ed89eda5. Apparently the add upload was not written for monolingual formats...
